src/core/reactor_backend.cc

```diff
@@ -1478,51 +1495,143 @@ class reactor_backend_uring final : public reactor_backend {
+                    }
+                    _result.set_value(bytes);
+                    delete this;
+                }
+                void set_exception(std::exception_ptr eptr) noexcept final {
+                    _result.set_exception(eptr);
+                    delete this;
+                }
+                future<size_t> get_future() {
+                    return _result.get_future();
+                }
+            };
+            auto desc = std::make_unique<read_completion>(fd, len);
+            auto req = internal::io_request::make_read(fd.fd.get(), -1, buffer, len, false);
+            auto fut = desc->get_future();
+            _r._io_sink.submit(desc.release(), std::move(req));
+            return fut;
         });
     }
     virtual future<size_t> read_some(pollable_fd_state& fd, const std::vector<iovec>& iov) override {
-        return readable(fd).then([this, &fd, iov = iov] () mutable {
+        if (is_speculated(fd, POLLIN)) {
             ::msghdr mh = {};
-            mh.msg_iov = &iov[0];
+            mh.msg_iov = const_cast<iovec*>(iov.data());
             mh.msg_iovlen = iov.size();
-            auto r = fd.fd.recvmsg(&mh, 0);
-            if (!r) {
-                return read_some(fd, iov);
+            auto r = fd.fd.recvmsg(&mh, MSG_DONTWAIT);
+            if (r) {
+                if (size_t(*r) == internal::iovec_len(iov)) {
+                    fd.speculate_epoll(EPOLLIN);
+                }
+                return make_ready_future<size_t>(*r);
             }
-            if (size_t(*r) == internal::iovec_len(iov)) {
-                fd.speculate_epoll(EPOLLIN);
+        }
+        class read_completion final : public io_completion {
+            pollable_fd_state& _fd;
+            std::vector<iovec> _iov;
+            ::msghdr _mh = {};
+            promise<size_t> _result;
+        public:
+            read_completion(pollable_fd_state& fd, const std::vector<iovec>& iov)
+                : _fd(fd), _iov(iov) {
+                _mh.msg_iov = const_cast<iovec*>(_iov.data());
+                _mh.msg_iovlen = _iov.size();
             }
-            return make_ready_future<size_t>(*r);
-        });
+            void complete(size_t bytes) noexcept final {
+                if (bytes == internal::iovec_len(_iov)) {
+                    _fd.speculate_epoll(EPOLLIN);
+                }
+                _result.set_value(bytes);
+                delete this;
+            }
+            void set_exception(std::exception_ptr eptr) noexcept final {
+                _result.set_exception(eptr);
+                delete this;
+            }
+            ::msghdr* msghdr() {
+                return &_mh;
+            }
+            future<size_t> get_future() {
+                return _result.get_future();
+            }
+        };
+        auto desc = std::make_unique<read_completion>(fd, iov);
+        auto req = internal::io_request::make_recvmsg(fd.fd.get(), desc->msghdr(), MSG_DONTWAIT);
```

Comment:
MSG_DONTWAIT is wrong here, we do want to wait, otherwise we can get EAGAIN.

Not sure how this worked for you, since http should never speculate as readable. Maybe I don't understand how MSG_DONTWAIT interacts with io_uring correctly?

Reply:
ahh. right! the uring call should be *blocking*. it was a silly replace-them-all bug. let me retest.

Reply:
> Not sure how this worked for you, since http should never speculate as readable. Maybe I don't understand how MSG_DONTWAIT interacts with io_uring correctly?

i came across this topic the other day. my understanding is that io_uring does return EAGAIN if the fd is opened with `O_NONBLOCK`. and `MSG_DONTWAIT` works like we expect. see https://patchwork.kernel.org/project/linux-fsdevel/patch/20190629161204.27226-2-axboe@kernel.dk/ .

guess i was just lucky when testing my change, as both the server (httpd) and the client (wrk) are located on the same machine.

Reply:
Yes, we should open the file without SOCK_NONBLOCK if uring sendmsg is available (and carefully make sure we always use SOCK_NONBLOCK everywhere).

Reply:
@avikivity i was wrong! i realized that io_uring *never* returns -EAGAIN when reading the commit message of https://github.com/torvalds/linux/commit/d7718a9d25a61442da8ee8aeeff6a0097f0ccfd6 .

so we don't need to care about the `SOCK_NONBLOCK` anymore. io_uring takes care of this already. and it is even more efficient after linux v5.7 !

Reply:
this also explains why `internal::io_request::make_recvmsg(fd.fd.get(), desc->msghdr(), MSG_DONTWAIT)` worked last week end.

Reply:
That's very convenient for us. Please ask it to be documented on the uring mailing list, this is not intuitive and in order to trust it not to change it's better if it's documented.